### ISSUE TITLE
[Task]: Why is the EfficientAD training so slow, run 200 epoch

### ISSUE THREAD

**Issue by kikyostudypython** [openvinotoolkit/anomalib#1164]
### What is the motivation for this task?

and what can I do to improve the inference speed, the verification section shows only 2.6fps, I want to improve it to 60-100fps, looking forward to your answer, thank you

### Describe the solution you'd like
```yaml
dataset:
  name: mvtec
  format: mvtec
  path: ./datasets/MVTec
  category: bottle
  task: segmentation
  train_batch_size: 1
  eval_batch_size: 16
  num_workers: 0
  image_size: 256 # dimensions to which images are resized (mandatory)
  center_crop: null # dimensions to which images are center-cropped after resizing (optional)
  normalization: none # data distribution to which the images will be normalized: [none, imagenet]
  transform_config:
    train: null
    eval: null
  test_split_mode: from_dir # options: [from_dir, synthetic]
  test_split_ratio: 0.2 # fraction of train images held out testing (usage depends on test_split_mode)
  val_split_mode: same_as_test # options: [same_as_test, from_test, synthetic]
  val_split_ratio: 0.5 # fraction of train/test images held out for validation (usage depends on val_split_mode)

model:
  name: efficientad
  teacher_out_channels: 384
  model_size: small # options: [small, medium]
  lr: 0.0001
  weight_decay: 0.00001
  padding: false
  pad_maps: true # relevant for "padding: false", see EfficientAD in lightning_model.py
  # generic params
  normalization_method: min_max # options: [null, min_max, cdf]

metrics:
  image:
    - F1Score
    - AUROC
  pixel:
    - F1Score
    - AUROC
  threshold:
    method: adaptive #options: [adaptive, manual]
    manual_image: null
    manual_pixel: null

visualization:
  show_images: False # show images on the screen
  save_images: True # save images to the file system
  log_images: False # log images to the available loggers (if any)
  image_save_path: null # path to which images will be saved
  mode: full # options: ["full", "simple"]

project:
  seed: 42
  path: ./results

logging:
  logger: [] # options: [comet, tensorboard, wandb, csv] or combinations.
  log_graph: false # Logs the model graph to respective logger.

optimization:
  export_mode: null # options: torch, onnx, openvino
# PL Trainer Args. Don't add extra parameter here.
trainer:
  enable_checkpointing: true
  default_root_dir: null
  gradient_clip_val: 0
  gradient_clip_algorithm: norm
  num_nodes: 1
  devices: 1
  enable_progress_bar: true
  overfit_batches: 0.0
  track_grad_norm: -1
  check_val_every_n_epoch: 1
  fast_dev_run: false
  accumulate_grad_batches: 1
  max_epochs: 200
  min_epochs: null
  max_steps: 70000
  min_steps: null
  max_time: null
  limit_train_batches: 1.0
  limit_val_batches: 1.0
  limit_test_batches: 1.0
  limit_predict_batches: 1.0
  val_check_interval: 1.0
  log_every_n_steps: 50
  accelerator: auto # <"cpu", "gpu", "tpu", "ipu", "hpu", "auto">
  strategy: null
  sync_batchnorm: false
  precision: 32
  enable_model_summary: true
  num_sanity_val_steps: 0
  profiler: null
  benchmark: false
  deterministic: false
  reload_dataloaders_every_n_epochs: 0
  auto_lr_find: false
  replace_sampler_ddp: true
  detect_anomaly: false
  auto_scale_batch_size: false
  plugins: null
  move_metrics_to_cpu: false
  multiple_trainloader_mode: max_size_cycle
```

### Additional context

```bash
PS D:\Project\anomalib-main> python tools/train.py --model efficientad
D:\Project\anomalib-main\anomalib\config\config.py:275: UserWarning: config.project.unique_dir is set to False. This does not ensure that your results will be written in an empty directory and you may overwrite files.
  warn(                                                                                                                                                                                                                  
Global seed set to 42                                                                                                       
2023-07-04 15:26:07,532 - anomalib.data - INFO - Loading the datamodule                                                     
2023-07-04 15:26:07,532 - anomalib.data.utils.transform - INFO - No config file has been provided. Using default transforms.
2023-07-04 15:26:07,532 - anomalib.data.utils.transform - INFO - No config file has been provided. Using default transforms.
2023-07-04 15:26:07,532 - anomalib.models - INFO - Loading the model.                                                       
2023-07-04 15:26:07,548 - anomalib.models.components.base.anomaly_module - INFO - Initializing EfficientadLightning model.
E:\AnacondaInstall\envs\anomalib_env\lib\site-packages\torchmetrics\utilities\prints.py:36: UserWarning: Metric `PrecisionRecallCurve` will save all targets and predictions in buffer. For large datasets this may lead to large memory fo
otprint.
  warnings.warn(*args, **kwargs)
2023-07-04 15:26:07,595 - anomalib.models.efficientad.lightning_model - INFO - Load pretrained teacher model from pre_trained\efficientad_pretrained_weights\pretrained_teacher_small.pth
2023-07-04 15:26:07,673 - anomalib.utils.loggers - INFO - Loading the experiment logger(s)
2023-07-04 15:26:07,673 - anomalib.utils.callbacks - INFO - Loading the callbacks
D:\Project\anomalib-main\anomalib\utils\callbacks\__init__.py:142: UserWarning: Export option: None not found. Defaulting to no model export
  warnings.warn(f"Export option: {config.optimization.export_mode} not found. Defaulting to no model export")
2023-07-04 15:26:07,751 - pytorch_lightning.utilities.rank_zero - INFO - GPU available: True (cuda), used: True
2023-07-04 15:26:07,751 - pytorch_lightning.utilities.rank_zero - INFO - TPU available: False, using: 0 TPU cores
2023-07-04 15:26:07,751 - pytorch_lightning.utilities.rank_zero - INFO - IPU available: False, using: 0 IPUs
2023-07-04 15:26:07,751 - pytorch_lightning.utilities.rank_zero - INFO - HPU available: False, using: 0 HPUs
2023-07-04 15:26:07,751 - pytorch_lightning.utilities.rank_zero - INFO - `Trainer(limit_train_batches=1.0)` was configured so 100% of the batches per epoch will be used..
2023-07-04 15:26:07,767 - pytorch_lightning.utilities.rank_zero - INFO - `Trainer(limit_val_batches=1.0)` was configured so 100% of the batches will be used..
2023-07-04 15:26:07,767 - pytorch_lightning.utilities.rank_zero - INFO - `Trainer(limit_test_batches=1.0)` was configured so 100% of the batches will be used..
2023-07-04 15:26:07,767 - pytorch_lightning.utilities.rank_zero - INFO - `Trainer(limit_predict_batches=1.0)` was configured so 100% of the batches will be used..
2023-07-04 15:26:07,767 - pytorch_lightning.utilities.rank_zero - INFO - `Trainer(val_check_interval=1.0)` was configured so validation will run at the end of the training epoch..
2023-07-04 15:26:07,767 - anomalib - INFO - Training the model.
2023-07-04 15:26:07,767 - anomalib.data.mvtec - INFO - Found the dataset.....
2023-07-04 15:26:07,767 - pytorch_lightning.utilities.rank_zero - INFO - You are using a CUDA device ('NVIDIA GeForce RTX 3060') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('mediu
m' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision
E:\AnacondaInstall\envs\anomalib_env\lib\site-packages\torchmetrics\utilities\prints.py:36: UserWarning: Metric `ROC` will save all targets and predictions in buffer. For large datasets this may lead to large memory footprint.
  warnings.warn(*args, **kwargs)
2023-07-04 15:26:07,923 - pytorch_lightning.accelerators.cuda - INFO - LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
2023-07-04 15:26:07,923 - pytorch_lightning.callbacks.model_summary - INFO - 
  | Name                  | Type                     | Params
-------------------------------------------------------------------
0 | image_threshold       | AnomalyScoreThreshold    | 0
1 | pixel_threshold       | AnomalyScoreThreshold    | 0
2 | model                 | EfficientADModel         | 8.1 M
3 | image_metrics         | AnomalibMetricCollection | 0
4 | pixel_metrics         | AnomalibMetricCollection | 0
5 | normalization_metrics | MinMax                   | 0
-------------------------------------------------------------------
8.1 M     Trainable params
0         Non-trainable params
8.1 M     Total params
32.235    Total estimated model params size (MB)
E:\AnacondaInstall\envs\anomalib_env\lib\site-packages\pytorch_lightning\trainer\connectors\data_connector.py:224: PossibleUserWarning: The dataloader, train_dataloader, does not have many workers which may be a bottleneck. Consider in
creasing the value of the `num_workers` argument` (try 8 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
E:\AnacondaInstall\envs\anomalib_env\lib\site-packages\pytorch_lightning\trainer\connectors\data_connector.py:224: PossibleUserWarning: The dataloader, val_dataloader 0, does not have many workers which may be a bottleneck. Consider in
creasing the value of the `num_workers` argument` (try 8 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
Training: 0it [00:00, ?it/s]2023-07-04 15:26:08,845 - anomalib.models.efficientad.lightning_model - INFO - Calculate teacher channel mean and std
Calculate teacher channel mean: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 209/209 [00:08<00:00, 24.50it/s] 
Calculate teacher channel std: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 209/209 [00:00<00:00, 13380.08it/s] 
Epoch 0:   0%|                                                                                                                                                                                                    | 0/215 [00:00<?, ?it/s]E
:\AnacondaInstall\envs\anomalib_env\lib\site-packages\pytorch_lightning\core\module.py:481: UserWarning: You called `self.log('train_st', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=ALo
gger(...))`
  rank_zero_warn(
E:\AnacondaInstall\envs\anomalib_env\lib\site-packages\pytorch_lightning\utilities\data.py:84: UserWarning: Trying to infer the `batch_size` from an ambiguous collection. The batch size we found is 1. To avoid any miscalculations, use 
`self.log(..., batch_size=batch_size)`.
  warning_cache.warn(
E:\AnacondaInstall\envs\anomalib_env\lib\site-packages\pytorch_lightning\core\module.py:481: UserWarning: You called `self.log('train_ae', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=AL
ogger(...))`
  rank_zero_warn(
E:\AnacondaInstall\envs\anomalib_env\lib\site-packages\pytorch_lightning\core\module.py:481: UserWarning: You called `self.log('train_stae', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=
ALogger(...))`
  rank_zero_warn(
E:\AnacondaInstall\envs\anomalib_env\lib\site-packages\pytorch_lightning\core\module.py:481: UserWarning: You called `self.log('train_loss', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=
ALogger(...))`
  rank_zero_warn(
Epoch 199:  97%|▉| 209/215 [002023-07-04 16:46:42,576 - anomalib.models.efficientad.lightning_model - INFO - Calculate Validation Dataset Quantiles.834, pixel_F1Score=0.791, pixel_AUROC=0.983, train_st_epoch=0.621, train_ae_epoch=0.14
Calculate Validation Dataset Quantiles: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6/6 [00:01<00:00,  3.40it/s]
Epoch 199: 100%|█| 215/215 [00:24<00:00,  8.62it/s, loss=0.801, train_st_step=0.628, train_ae_step=0.174, train_stae_step=0.0326, train_loss_step=0.834, pixel_F1Score=0.795, pixel_AUROC=0.983, train_st_epoch=0.620, train_ae_epoch=0.142
023-07-04 16:46:48,193 - pytorch_lightning.utilities.rank_zero - INFO - `Trainer.fit` stopped: `max_epochs=200` reached.
Epoch 199: 100%|█| 215/215 [00:25<00:00,  8.46it/s, loss=0.801, train_st_step=0.628, train_ae_step=0.174, train_stae_step=0.0326, train_loss_step=0.834, pixel_F1Score=0.795, pixel_AUROC=0.983, train_st_epoch=0.620, train_ae_epoch=0.14
2023-07-04 16:46:48,284 - anomalib.utils.callbacks.timer - INFO - Training took 4840.36 seconds
2023-07-04 16:46:48,284 - anomalib - INFO - Loading the best model weights.
2023-07-04 16:46:48,285 - anomalib - INFO - Testing the model.
2023-07-04 16:46:48,290 - anomalib.data.mvtec - INFO - Found the dataset.....
2023-07-04 16:46:48,291 - pytorch_lightning.utilities.rank_zero - INFO - You are using a CUDA device ('NVIDIA GeForce RTX 3060') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('mediu
m' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision
2023-07-04 16:46:48,291 - anomalib.utils.callbacks.model_loader - INFO - Loading the model from D:\Project\anomalib-main\results\efficientad\mvtec\bottle\run\weights\model-v1.ckpt
2023-07-04 16:46:48,388 - pytorch_lightning.accelerators.cuda - INFO - LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
E:\AnacondaInstall\envs\anomalib_env\lib\site-packages\pytorch_lightning\trainer\connectors\data_connector.py:224: PossibleUserWarning: The dataloader, test_dataloader 0, does not have many workers which may be a bottleneck. Consider i
ncreasing the value of the `num_workers` argument` (try 8 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
Testing DataLoader 0: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6/6 [00:31<00:00,  5.24s/it]2
023-07-04 16:47:20,295 - anomalib.utils.callbacks.timer - INFO - Testing took 31.876209020614624 seconds
Testing DataLoader 0: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6/6 [00:31<00:00,  5.32s/it] 
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       Test metric             DataLoader 0
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       image_AUROC                  1.0
      image_F1Score                 1.0
       pixel_AUROC          0.9827045202255249
      pixel_F1Score         0.7951030135154724
```

**Comment by alexriedel1** [comment]
To measure the timing of the model you shouldn't use the pytorch lightning testing function I guess. There is a lot of overload, data loading, model loading whatever included probably. Just write a clean inferencing script, be sure to synchronize cuda, do some warmup steps and find out how to measure time correctly in pytorch https://leimao.github.io/blog/PyTorch-Benchmark/

**Comment by Kyle-fang** [comment]
Hello! I'm experiencing the same problem as you (i.e. very slow reasoning, which doesn't match the official data), have you found the cause? And if there is a solution? @kikyostudypython @alexriedel1

**Comment by alexriedel1** [comment]
> Hello! I'm experiencing the same problem as you (i.e. very slow reasoning, which doesn't match the official data), have you found the cause? And if there is a solution? @kikyostudypython @alexriedel1

Please see here for measuring the timing: https://github.com/openvinotoolkit/anomalib/discussions/1183

**Comment by samet-akcay** [comment]
Hi @kikyostudypython and @Kyle-fang,  as @alexriedel1 pointed out, `LightningInferencer` runs extra stuff, which overall impacts the inference speed. 

I'm closing this issue as it seems duplicate of #1183.
